README
                                 ART
                       Allotment Routing Table

                           Yoichi Hariguchi




1. Contents

This package includes two ART implementations:

  o ipArt: supports arbitrary address length and arbitrary
    stride length distribution. `ipArt' uses simple multibit
    trie.
  o ipArt-PC supports arbitrary address length and arbitrary
    stride length distribution. `ipArt-PC' uses path-compressed
    trie.

`ipArt' and `ipArt-PC' are stored in the same library (libipart.a.)
You can choose which one is used when instantiating a routing
table.



2. Files

  Makefile              Makefile
  README                This file
  test.sh               Runs addition, longest prefix match, exact match,
                        and deletion tests as follows:
                          1. Simple trie with 569,770 IPv4 prefixes
                          2. Path-compressed trie with 569,770 IPv4 prefixes
                          3. Simple trie with 24,470 IPv6 prefixes
                          4. Path-compressed trie with 24,470 IPv6 prefixes
  Types.h               Local type definitions
  ipArt.c               An `ipArt' ART implementation (simple trie)
  ipArtPathComp.c       An `ipArt-PC' ART implementation
                        (path-compressed trie)
  ipArt.h               Header file of `ipArt' and `ipArt-PC'
  util.c                utility functions (obsolete)
  data/
   v4routes-random1.txt 569,770 IPv4 prefixes in random order
                        for the route insertion test
   v4routes-random2.txt 569,770 IPv4 prefixes in random order
                        for the route search test
   v4routes-random3.txt 569,770 IPv4 prefixes in the random
                        order for the route deletion test
   v6routes-random1.txt 24,470 IPv6 prefixes in random order
                        for the route insertion and search test
   v6routes-random2.txt 24,470 IPv6 prefixes in random order
                        for the route deletion and search test

Note: v{4,6}routes-random*.txt are created from the Jul 7, 2015
output of 'ip route' and 'ipv6 route' respectively at
telnet://route-views.routeviews.org.


3. Installation

Type `make' and a test command `rtLoookup' and library
`libipart.a' are built.


4. Interactive Simulation

  - IPv4 simple trie:          ./rtLookup 4 simple 
  - IPv4 path-compressed trie: ./rtLookup 4 pc
  - IPv6 simple trie:          ./rtLookup 6 simple 
  - IPv6 path-compressed trie: ./rtLookup 6 pc


5. Data Structures

5.1. rtTable: Routing Table

`rtTable' is used to represent a routing table. It is not
necessary for users to access the members of `rtTable'.


5.2. routeEnt: Route Entry

The definition of `routeEnt' in `ipArt' and `ipArt-PC' is as
follows:

  typedef struct routeEnt {
      u8  dest[16];               /* upto IPv6 */
      int plen;                   /* prefix length */
      int level;                  /* for performance */
      /*
       * Your stuff
       */
  } routeEnt, *pRouteEnt;

`dest' must represent the destination address of the route.
`plen' must represent the corresponding prefix length.
The destination address must be stored in the network byte
order.


6. API functions

6.1. Routing Table Creation

rtTable *
rtArtInit (int nLevels, s8* psl, int alen, trieType type)

 @name  rtArtInit

 @brief API Function.
        Creates and initializes a routing table.
        Example usage:
          - IPv4 routing table
          - The stride lengths are:
             level 0: 16 bits
             level 1:  8 bits
             level 2:  8 bits
          - Use a path-compressed trie

         s8      sl[3] = { 16, 8, 8 };
         rtTable pt    = rtArtInit(3, sl, 32, pathCompTrie)

 @param[in] nLevels The number of trie node levels
                    (or the number of stride lengths)
 @param[in] psl     Pointer to an array of stride lengths
 @param[in] alen    Bit length of IP addresses (32 or 128)
 @param[in] type    simpleTrie (0) or pathCompTrie (1).

 @retval rtTable* Pointer to the allocated routing table
 @retval NULL     Failed to allocate a new routing table

Hereafter, let the routing table pointer be:

  rtTable* pt;   /* Pointer to the routing table */


6.2. Memory Allocation for a New Route

routeEnt*
rtArtNewRoute(rtTable* pt)

 @brief  API function.
         Allocates memory for a new route entry.
 
 @param[in] pt Pointer to the routing table
 
 @retval routeEnt* Pointer to the newly allocated route.
 @retval NULL      There was no memory for a new route.


6.3. Insertion

routeEnt*
pt->insert(rtTable pt, routeEnt* pEnt)

 @brief  API function.
         Add a route represented by `pEnt' to the routing table `pt'

 @param[in] pt   Pointer to the routing table
 @param[in] pEnt Pointer to the route added to `pt'.
                 `pEnt' must NOT point to a local variable.

 @retval routeEnt* `pEnt' is successfully inserted.
         pEnt      There is an existing route that has the same
                   IP prefix (address and prefix length).
                   `pEnt' must be freed in this case.


6.4. Deletion

bool
pt->delete(rtTable pt, u8* dest, int plen)

 @brief  API function.
         Deletes a route represented by an IP prefix (
         (address and its prefix length) from the routing table.
         The matched route entry is freed in this function.

 @param[in] pt    Pointer to the routing table
 @param[in] pDest Pointer to the IP address to be deleted from `pt'
 @param[in] plen  Prefix length associated with `pDest'

 @retval ture  If the matching route is deleted from `pt'.
               The route entry is freed in this function.
 @retval false If there is no matching route in `pt'.


6.5. Longest Prefix Match

routeEnt*
pt->findMatch(rtTable* pt, u8* pDest)

 @brief  API function.
         Perform the longest prefix match.

 @param[in] pt          Pointer to the routing table
 @param[in] pDest       Pointer to the destination IP address

 @retval routeEnt* Pointer to the longest prefix matching route.
 @retval NULL      There was no matching route for `pDest'


6.6. Exact Route Match

routeEnt *
pt->findExactMatch(rtTable* pt, u8* pDest, int plen)

 @brief API Function.
        Performs the exact match (address + prefix length.)

 @param[in] pt    Pointer to the routing table
 @param[in] pDest Pointer to the IP address to be searched for
 @param[in] plen  prefix length of `pDest'

 @retval routeEnt* Pointer to the found route entry (success)
 @retval NULL      Failed to find a matching route entry


6.7. Delete All Routes

pt->flush(rtTable* pt)

 @brief API Function.
        Delete all the route entries in the routing table.
 @param[in] pt Pointer to the routing table


6.8. Delete the Routing Table

 void pt->deleteTable(rtTable** p)

 @brief API Function.
        Delete all the route entries in the routing table, then
        Free the routing table itself.
 @param[in,out] p Pointer to the pointer to `rtTable'. `*p' is
                  set to NULL at the end of this function.


Changelog
2001-06-12  1.0  Released
2001-09-15  1.1  Fixed bug that wrongly reports that the
                 destination IP address is not included in the
                 matched prefix
2001-12-30  1.2  Fixed bug: init value of `i' in rtArtPcDeleteRoute()
2015-07-05  1.3  Added a lot of comments. Started moving the
                 code to GitHub
2015-07-09  1.4  Added exact route match function.
2015-07-10  1.5  Fixed bugs in rtArtPcFindMatch(). It is
                 necessary to do bit string comparison in the
                 case of path compressed trie (no need to do so
                 in the case of simple trie.)
2015-07-10  1.6  Modified IPv6 test to support up to /128 routes.
2015-07-17  1.7  Added support for >8 bit stride length subtables to
                 path-compressed trie
2016-08-22  1.8  Added the following features:
                  1. Flush Routes: This feature deletes all the
                     routes from the routing table.
                  2. Delete Routing Table: This feature first
                     delete all the routes by calling flush
                     route function, then delete the routing
                     table itself and free its memory.


tests.sh
#!/bin/sh
#

echo "1. IPv4 simple tire"
./rtLookup 4 simple perf

echo && echo && echo "2. IPv4 path-compressed tire:"
./rtLookup 4 pc perf

echo && echo && echo "3. IPv6 simple tire:"
./rtLookup 6 simple perf

echo && echo && echo "4. IPv6 path-compressed tire:"
./rtLookup 6 pc perf


ipArt.h
/*
   Copyright (c) 2001-2016
   Yoichi Hariguchi. All rights reserved.

   The algorithm of ART is invented by Donald Knuth in 2000 while he was
   reviewing Yoichi's paper (http://www.hariguchi.org/art/smart.pdf).
   This is an ART implementation by Yoichi Hariguchi that supports
   arbitrary address length and multiple routing tables.
   This implementation includes a small part of Don Knuth's original ART
   implementation for IPv4.

   Permission to use, copy, modify, and distribute this software for any
   purpose with or without fee is hereby granted, provided that the above
   copyright notice and this permission notice appear in all copies.

   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/


#ifndef __ipArt_h__
#define __ipArt_h__
#ifdef __cplusplus
extern "C" {
#endif


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "Types.h"
#include "dllist.h"



typedef enum {
    simpleTrie   = 0,
    pathCompTrie = 1,           /* trie with path compression */
} trieType;

typedef struct routeEnt {
    u8  dest[16];               /* upto IPv6 */
    int plen;                   /* prefix length */
    int level;                  /* for performance */
} routeEnt, *pRouteEnt;


typedef union tableEntry {
    routeEnt         *ent;
    union tableEntry *down;
    union {
        struct {
            s16 nRoutes;     /* number of routes in this subtable */
            s16 nSubtables;  /* number of subtabls in this subtable */
        };
        size_t count;           /* reference counter */
    };
    size_t level;               /* t[-1] holds level */
} tableEntry;
typedef tableEntry *subtable;


typedef struct {
    u8 sb;                  /* start byte from the beg. of address */
    u8 bo;                  /* bit offset from bit 7 */
    u8 sl;                  /* stride length */
    u8 tl;                  /* total stride length up to this level */
} strideInfo;


typedef struct pcSubtbls pcSubtbls;
struct pcSubtbls {
    subtable pst;
    int      idx;
};

typedef struct rtTable rtTable;
struct rtTable {
    tableEntry*  root;    /* pointer to root subtable */
    strideInfo*  psi;     /* array of stride information indexed by level */
    union {
        tableEntry** pEnt;/* array of table entry pointers */
        routeEnt**   pDef;/* array of subtable (trie node) default routes */
    };
    subtable*    pTbl;    /* array of subtable pointers (for deletion) */
    pcSubtbls*   pPcSt;   /* array of pcSubtables pointers (for deleteion) */
    s16          alen;    /* address length in bits for search */
    u16          len;     /* address length in bytes */
    s16          off;     /* -1 - byte2nPtrs(len) */
    u16          nLevels; /* number of levels */

    routeEnt* (*insert)(rtTable* p, routeEnt* r);
    bool (*delete)(rtTable* p, u8* pDest, int plen);
    bool (*flush)(rtTable* pt);
    void (*deleteTable)(rtTable** pt);
    routeEnt* (*findMatch)(rtTable *p, u8* pDest);
    routeEnt* (*findExactMatch)(rtTable *p, u8* pDest, int plen);

    int  nRoutes;           /* # of routes */
    int* nHeaps;            /* # of heaps at level `i' */
    int* nTransit;          /* # of transit heaps at level `i'  */
    u32  nSubtablesFreed;   /* # of freed subtables (for debugging) */
};

typedef struct rtArtWalkQnode rtArtWalkQnode;
struct rtArtWalkQnode
{
    dllNode  dlln;              /* doubly linked list node */
    subtable p;
};

typedef void (*rtFunc)(routeEnt*, void*);


#define isSubtable(p) ((p).count&1)
#define makeSubtable(p) (subtable)((size_t)(p)|1)
#define subtablePtr(p) ((tableEntry)((p).count&-2))


/*
 * API functions
 */
routeEnt* rtArtNewRoute(rtTable *pt);
void      rtArtFreeRoute(rtTable *pt, routeEnt *);
rtTable*  rtArtInit(int nLevels, s8* psl, int alen, trieType type);
rtTable*  rtArtPcInit(rtTable* pt);
bool      rtArtFlushRoutes(rtTable* pt);
void      rtArtWalkTable(rtTable* pt, subtable p, int index,
                         int thresh, rtFunc f, void* p2);
void      rtArtBFwalk(rtTable* pt, subtable p, rtFunc f, void* p2);
void      rtArtDFwalk(rtTable* pt, subtable p, rtFunc f, void* p2);
void      rtArtCollectStats(rtTable* pt, subtable ps);


/*
 * Inline functions
 */

/**
 * @name  bitCmp8
 *
 * @brief Compare between arbitrary bits of a byte
 *        
 *
 * @param[in] d1  Data 1 (one byte) to be compared.
 * @param[in] d2  Data 2 (one byte) to be compared.
 * @param[in] st  First bit (0-7) to be compared.
 * @param[in] end Last bit (0-7) to be compared.
 *
 * @retval Negative integer If bit st:end of d1 < bit st:end of d2
 * @retval 0                If bit st:end of d1 == bit st:end of d2
 * @retval Positive integer If bit st:end d1 > bit st:end of d2
 */
static inline int
bitCmp8 (u8 p1, u8 p2, int st, int end)
{
    u8 mask;

    assert(end >= st);
    assert(st >= 0 && st < 8);
    assert(end >= 0 && end < 8);

    mask =  (((u8)(~0)) >> (8 - (end - st + 1))) << st;
    p1 &= mask;
    p2 &= mask;
    return (p1 - p2);
}

/**
 * @name  bitStrCmp
 *
 * @brief Compare two bit strings
 *        
 *
 * @param[in] p1  Pointer to bit string 1 to be compared.
 * @param[in] p2  Pointer to bit string 2 to be compared.
 * @param[in] st  Start bit to be compared (bit 0 is the 7th bit of p1[0].)
 * @param[in] end End bit to be compared (bit 0 is the 7th bit of p1[0].)
 * @param[out] idx Pointer to the array index where the comparison
 *                 finished. This pointer can be NULL.
 *
 * @retval Negative integer If st:end bits of p1 < st:end bits of p2
 * @retval 0                If st:end bits of p1 == st:end bits of p2
 * @retval Positive integer st:end bits of p1 > st:end bits of p2
 */
static inline int
bitStrCmp (u8* p1,u8* p2, int st, int end, int* idx)
{
    register int sb;            /* start bit */
    int *i;
    int ii;
    int rc;

    assert(st <= end);

    i = (idx) ? idx : &ii;

    /* compare p1[*i] and p2[*i]
     */
    *i = st >> 3;
    sb = *i << 3;
    st -= sb;
    st = 7 - st;                /* end bit */
    if (end < (sb + 8)) {
        end -= sb;
        end = 7 - end;          /* start bit */
        return bitCmp8(p1[*i], p2[*i], end, st);
    } else {
        /*
         * more bits to compare
         */
        rc = bitCmp8(p1[*i], p2[*i], 0, st);
        if (rc) {
            return rc;
        }
        for (++*i, sb += 8; end >= (sb + 8); ++*i, sb += 8) {
            if (p1[*i] != p2[*i]) {
                return (p1[*i] - p2[*i]);
            }
        }
    }
    st = 7 - (end - sb);
    return bitCmp8(p1[*i], p2[*i], st, 7);
}


/**
 * @name  bytes2nPtrs
 *
 * @brief Convert number of bytes to the number of pointers
 *        This function works both on 32-bit and 64-bit CPUs.
 *
 * @param[in] nBytes Number of bytes
 *
 * @retval u32 The minimum number of pointers to cover `nBytes'
 */
static inline u32
bytes2nPtrs (u32 nBytes)
{
#if __SIZEOF_POINTER__ == 8
    return ((nBytes & 7) == 0) ? (nBytes >> 3) : ((nBytes >> 3) + 1);
#elif __SIZEOF_POINTER__ == 4
    return ((nBytes & 3) == 0) ? (nBytes >> 2) : ((nBytes >> 2) + 1);
#else
#error Pointer size must be 4 or 8.
#endif
}


/**
 * @name  bits2bytes
 *
 * @brief Convert number of bits to the number of bytes
 *
 * @param[in] nBits Number of bits
 *
 * @retval u32 The minimum number of bytes to cover `nBits'
 */
static inline u32
bits2bytes (u32 nBits)
{
    register u32 rv = nBits >> 3;
    if ((nBits & 7)) {
        return (rv + 1);
    }
    return rv;
}


/**
 * @name  cmpAddr
 *
 * @brief Compare two bit strings for the first 'plen' bits
 *
 * @param[in] p1   Pointer to bit string 1 to be compared.
 * @param[in] p2   Pointer to bit string 2 to be compared.
 * @param[in] plen Number of bits to be compared from the beginning
 *
 * @retval true  Two bit strings are identical
 * @retval false Two bit strings are different
 */
static inline bool
cmpAddr (u8* p1, u8* p2, int plen)
{
    int len;
    u8  mask;


    for ( len = 8; len <= plen; len += 8 ) {
        if ( *p1++ != *p2++ ) {
            return false;
        }
    }
    if ( len == plen ) {
        return true;
    }

    /*
     * check remining bits < 8
     */
    mask = ~((1 << (len - plen)) - 1);
    if ( (*p1 & mask) != (*p2 & mask) ) {
        return false;
    }
    return true;
}

/**
 * @name  plen2level
 *
 * @brief Finds the trie level of the given plefix length
 *
 * @param[in] pt   Pointer to the routing table
 * @param[in] plen prefix length
 *
 * @retval int The trie level of `plen'
 */
static inline int
plen2level (rtTable* pt, int plen)
{
    int l;


    l = 0;
    for (;;) {
        plen -= pt->psi[l].sl;
        if ( plen <= 0 ) break;
        ++l;
    }
    return l;
}


/**
 * @name  fringeIndex
 *
 * @brief Returns the fringe index.
 *          It takes `nBits' bits from bit `**p + *offset'
 *          and stores them to `c'.
 *        Side effect:
 *          `*p' and `*offset' are advanced after the process.
 *        Requirement:
 *          `nBits' must be <= 24, `*offset' must be < 8.
 *
 * @param[in]    nBits  Stride length to calculate the fringe index
 * @param[inout] p      Pointer to the pointer of the destination address
 * @param[inout] offset Pointer to the bit offset from '*p'
 *
 * @retval u32 The fringe index
 */
static inline u32
fringeIndex (u8** p, u32* offset, int nBits)
{
    u32 b, c;

    c = *offset;
    assert((c < 8) && (nBits <= 24));

    b = c + nBits;
    if ( b <= 8 ) {
        c = (**p >> (8 - b)) & ((1 << nBits) - 1);
    } else if ( b <= 16 ) {
        b -= 8;                 /* offset of byte *((*p)+1) */
        c  = ((*((*p)++) & ((1 << (8 - c)) - 1)) << b);
        c |= (**p >> (8 - b));
    } else if ( b <= 24 ) {
        b -= 16;                /* offset of byte *((*p)+2) */
        c  = ((*((*p)++) & ((1 << (8 - c)) - 1)) << (8 + b));
        c |= ((*((*p)++)) << b);
        c |= (**p >> (8 - b));
    } else {
        b -= 24;                    /* offset of byte *((*p)+3) */
        c  = ((*((*p)++) & ((1 << (8 - c)) - 1)) << (16 + b));
        c |= ((*((*p)++)) << (8 + b));
        c |= ((*((*p)++)) << b);
        c |= (**p >> (8 - b));
    }
    if ( b == 8 ) {
        b = 0;
        (*p)++;
    }
    *offset = b;
    return c + (1 << nBits);
}


/**
 * @name   baseIndex
 *
 * @brief Returns the base index
 *
 * @param[in] pTbl  Pointer to the routing table
 * @param[in] p     Poiner to the destination address of the route
 * @param[in] oplen Prefix length of the route
 *
 * @retval int The base index
 */
static inline int
baseIndex (rtTable* pTbl, u8* p, int plen)
{
    register int len  = 0;
    register int sl;            /* stride length at level `i' */
    register int i;
    u32 st;                     /* stride */

    for ( i = 0;; ++i ) {
        assert(i < pTbl->nLevels);

        sl = pTbl->psi[i].sl;
        if ( plen <= len + sl ) break;
        len += sl;
    }
    p += (len >> 3);         /* sl/8 */

    plen = plen - len;          /* prefix length in stride */
    i = len & 7;                /* i = bit offset in first byte */
    len = i + sl;               /* len = offset + stride length */
    if ( len <= 8 ) {
        st  = (*p >> (8 - len)) & ((1 << sl) - 1);
    } else if ( len <= 16 ) {
        len -= 8;               /* offset in second byte */
        st = ((p[0] & ((1 << (8 - i)) - 1)) << len) | (p[1] >> (8 - len));
    } else if ( len <= 24 ) {
        len -= 16;              /* offset in third byte */
        st = ((p[0] & ((1 << (8 - i)) - 1)) << (8 + len)) |
            (p[1] << len) | (p[2] >> (8 - len));
    } else {
        assert(len <= 32);

        len -= 24;
        st = ((p[0] & ((1 << (8 - i)) - 1)) << (16 + len)) |
            (p[1] << (8 + len)) | (p[2] << len) | (p[3] >> (8 - len));
    }

    return (st >> (sl - plen)) + (1 << plen);
}


/**
 * @name   rtArtAllot
 *
 * @brief  1. Starting from index `k' of subtable (trie node) `t',
 *            visits all the child indices of `k'.
 *         2. Replaces `r' with `s' if the visited indices has `r'
 *         3. Stop visiting when the visited index does not have `r'.
 *            This is because the rest of child indices are supposed to
 *            have more specific route.
 *         Note:
 *         This function was originally written by Donald Knuth in CWEB.
 *         The complicated `goto's are the result of expansion of
 *         two recursions.
 *
 * @param[in] t           Pointer to a subtable (trie node)
 * @param[in] k           Index to start process.
 *                        `k' must be smaller than `threshold'
 * @param[in] r           Route pointer to be replaced with 's'
 * @param[in] s           Route pointer to replace 'r'
 * @param[in] threshold   The first fringe index of 't'
 * @param[in] fringeCheck False if `t' is the deepest level. Otherwise true.
 *
 * @retval int The base index
 */
static inline void
rtArtAllot (subtable t, int k, routeEnt *r,
            routeEnt *s, int threshold, bool fringeCheck)
{
    register int j = k;

    assert( k < threshold );

startChange:
    j <<= 1;
    if ( j < threshold ) goto nonFringe;


    /*  change fringe nodes
     */
    while (1) {
        if ( fringeCheck && isSubtable(t[j]) ) {
            if ( subtablePtr(t[j]).down[1].ent == r ) {
                subtablePtr(t[j]).down[1].ent = s;
            }
        } else if ( t[j].ent == r ) {
            t[j].ent = s;
        }
        if ( j & 1 ) goto moveUp;
        j++;
    }

nonFringe:
    if (t[j].ent == r) goto startChange;
moveOn:
    if (j & 1) goto moveUp;
    j++;
    goto nonFringe;
moveUp:
    j >>= 1;
    t[j].ent = s;               /* change non-fringe node */
    if (j != k) goto moveOn;
}


#endif /* __ipArt_h__ */


dllist.h
#ifndef __DLLIST_H_
#define __DLLIST_H_

/* $Id: dllist.h,v 1.1 2015/06/25 21:19:27 cvsremote Exp $

   dllist.h: doubly linked list library


   Copyright (c) 2011, Yoichi Hariguchi
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are
   met:

       o Redistributions of source code must retain the above copyright
         notice, this list of conditions and the following disclaimer.
       o Redistributions in binary form must reproduce the above
         copyright notice, this list of conditions and the following
         disclaimer in the documentation and/or other materials provided
         with the distribution.
       o Neither the name of the Yoichi Hariguchi nor the names of its
         contributors may be used to endorse or promote products derived
         from this software without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 */

#ifndef OPTIMIZATION_ON
#pragma GCC diagnostic ignored "-Wunused-function"
#endif /* OPTIMIZATION_ON */


/*
 * Doubly linked list definitions
 */
typedef struct dllNode_ {
    struct dllNode_* next;
    struct dllNode_* prev;
} dllNode;
typedef struct dllHead_ {
    dllNode* head;
    dllNode* tail;
} dllHead;


static inline void
dllInit (dllHead* ph)
{
    ph->head = ph->tail = NULL;
}
static inline void
dllPushNode (dllHead* ph, dllNode* pn)
{
    if (!ph->head) {
        pn->next = pn->prev = NULL;
        ph->head = ph->tail = pn;
        return;
    }
    ph->head->prev = pn;
    pn->next = ph->head;
    pn->prev = NULL;
    ph->head = pn;
}
static inline dllNode*
dllPopNode (dllHead* ph)
{
    dllNode* pn;

    pn = ph->head;
    if (!pn) {
        return NULL;
    }
    if (pn == ph->tail) {
        ph->head = ph->tail = NULL;
        return pn;
    }
    ph->head = pn->next;
    ph->head->prev = NULL;
    return pn;
}
static inline void
dllEnqNode (dllHead* ph, dllNode* pn)
{
    if (!ph->tail) {
        pn->next = pn->prev = NULL;
        ph->head = ph->tail = pn;
        return;
    }
    pn->next       = NULL;
    pn->prev       = ph->tail;
    ph->tail->next = pn;
    ph->tail       = pn;
}
static inline dllNode*
dllRmLastNode (dllHead* ph)
{
    dllNode* pn;

    pn = ph->tail;
    if (!pn) {
        return NULL;
    }
    if (pn == ph->head) {
        ph->head = ph->tail = NULL;
        return pn;
    }
    ph->tail       = pn->prev;
    pn->prev->next = NULL;
    return pn;
}
static inline void
dllPrependNode (dllHead* ph, dllNode* pn, dllNode* pnNew)
{
    if (ph->head == pn) {
        return dllPushNode(ph, pnNew);
    }
    pnNew->next    = pn;
    pnNew->prev    = pn->prev;
    pn->prev->next = pnNew;
    pn->prev       = pnNew;
}
static inline void
dllAppendNode (dllHead* ph, dllNode* pn, dllNode* pnNew)
{
    if (ph->tail == pn) {
        return dllEnqNode(ph, pnNew);
    }
    pnNew->next    = pn->next;
    pnNew->prev    = pn;
    pn->next->prev = pnNew;
    pn->next       = pnNew;
}

static inline void
dllMergeLists (dllHead* ph1, dllHead* ph2)
{
    ph1->tail->next = ph2->head;
    ph1->tail       = ph2->tail;
    ph2->head       = NULL;
    ph2->tail       = NULL;
}

static inline void
dllSplitList (dllHead* ph1, dllHead* ph2, dllNode* pn)
{
    if (pn->prev == NULL) {
        ph2->head = pn;
        ph2->tail = ph1->tail;
        ph1->head = NULL;
        ph1->tail = NULL;
        return;
    }
    ph2->head = pn;
    ph2->tail = ph1->tail;
    ph1->tail = pn->prev;
    pn->prev  = NULL;
    ph1->tail->next = NULL;
}

#endif /* __DLLIST_H_ */


util.c
#include <stdlib.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "Types.h"


/* 1. buf must have 16 byte space
   2. add must be in the host byte order
 */
char *
inet_h2a(ipv4a add, char *buf)
{
    static char s[16];

    if ( !buf ) {
        buf = s;
    }

    sprintf(buf, "%d.%d.%d.%d",
            (add >> 24) & 0xff,
            (add >> 16) & 0xff,
            (add >> 8) & 0xff,
            add & 0xff);

    return buf;
}

/* 1. buf must have 16 byte space
   2. add must be in the network byte order
 */
char *
inet_n2a(ipv4na add, char *buf)
{
    static char s[16];

    if ( !buf ) {
        buf = s;
    }

    sprintf(buf, "%d.%d.%d.%d",
            add & 0xff,
            (add >> 8) & 0xff,
            (add >> 16) & 0xff,
            (add >> 24) & 0xff);

    return buf;
}


void
inet_a2n(int ver, char* s, u8* p)
{
    char* s2;
    ipv4a ipa;
    int   i;

    if ( ver == 4 ) {
        ipa = ntohl(inet_addr(s));
        p[0] = ipa >> 24;
        p[1] = (ipa >> 16) & 0xff;
        p[2] = (ipa >> 8) & 0xff;
        p[3] = ipa & 0xff;
        return;
    }
    s2 = s;
    ver = 1;                    /* user ver for flag */
    for ( i = 0; ver; ++i ) {
        for (;;) {
            if ( *s2 == '\0' ) {
                ver = 0;
                break;
            }
            if ( *s2 == ':' ) {
                *s2 = '\0';
                break;
            }
            ++s2;
        }
        p[i] = strtoul(s, NULL, 16);
        if ( ver ) {
            *s2 = ':';
        } else {
            return;
        }
        ++s2;
        s = s2;
    }
}


/* 1. buf must have 18 byte space
   2. add must be in the network byte order
 */
char *
inetStr(u8* add, int ver, char *buf)
{
    static char s[18];

    if ( !buf ) {
        buf = s;
    }

    switch ( ver ) {
    case 4:
        sprintf(buf, "%d.%d.%d.%d", add[0], add[1], add[2], add[3]);
        break;
    case 6:
        sprintf(buf, "%02x:%02x:%02x:%02x:%02x:%02x",
                add[0], add[1], add[2], add[3], add[4], add[5]);
        break;
    default:
        buf = NULL;
        break;
    }

    return buf;
}


static s8 magic[37] = {
     0,  5,  9, 30,  6, 31, 32,  0, 14, 13,
    24, 12, 27, 23, 18, 11,  0, 26, 20, 22,
     3, 17,  1, 10,  7,  0, 15, 25, 28, 19,
     0, 21,  4,  2,  8, 16, 29
};

int
mask2plen(ipv4a mask)
{
    return magic[mask % 37];
}
